core/blockchain.go

```diff
@@ -1672,6 +1672,15 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, er
 				"uncles", len(block.Uncles()), "txs", len(block.Transactions()), "gas", block.GasUsed(),
 				"root", block.Root())
 
+			// Special case. Commit the empty receipt slice if we meet the known
+			// block in the middle. It can only happen in the clique chain. Whenever
+			// we insert blocks via `insertSideChain`, we only commit `td`, `header`
+			// and `body` if it's non-existent. Since we don't have receipts without
+			// reexecution, so nothing to commit. But if the sidechain will be adpoted
+			// as the canonical chain eventually, it needs to be reexecuted for missing
+			// state, but if it's this special case here(skip reexecution) we will lose
+			// the empty receipt entry.
+			rawdb.WriteReceipts(bc.db, block.Hash(), block.NumberU64(), nil)
```

Comment:
Does not this presume that the block is empty? I don't think there's any guarantee about that..? 
Consider: 
```
block A: state X
block B1 (1tx): -> state Y  , block B2 (0 tx) -> state Z
block C1( 0tx):-> state O  , block C2 (1tx) -> state O
```

Reply:
Yes, let me think..

Reply:
Actually I think it's impossible. We are importing a batch of `known` blocks in `InsertChain`. The only use case is for regenerating the missing state for "side chain".

For the "side chain", they are checked that the relevant state is missing by `insertSideChain` function.

While in the special case: "so importing an older block might complete the state of the subsequent one", it's only possible in clique chain and the second block is empty.

Reply:
But I agree it's dangerous to commit the nil here. I will add a check, if the transaction list is empty, then commit the nil, otherwise print an error log. Do you think it makes sense?

Reply:
That makes sense. It's a bit difficult to reason about this, and make sure we cover all possibilities, so for me it sounds good to print out the error, so we can at least notice if our assumptions hold.